src/metabase/api/query_metadata.clj

```diff
@@ -15,30 +22,66 @@
     (catch Exception e
       (log/warnf "Error in dashboard metadata %s %s: %s" entity-type entity-id (ex-message e)))))
 
+(defn- partition-table-ids
+  [ids]
+  (reduce (fn [acc id]
+            (if-let [card-id (lib.util/legacy-string-table-id->card-id id)]
+              (update acc :card-ids conj card-id)
+              (update acc :table-ids conj id)))
+          {:card-ids [] :table-ids []}
+          ids))
+
+(defn- metadata-for-table-dependents
+  [table-dependents]
+  (let [{:keys [card-ids table-ids]} (partition-table-ids (map :id table-dependents))
+        card-metadatas (api.table/batch-fetch-card-query-metadatas card-ids)
+        table-metadatas (api.table/batch-fetch-table-query-metadatas table-ids)
+        all-metadatas (-> []
+                          (into (sort-by :id table-metadatas))
+                          (into (sort-by :id card-metadatas)))]
+    (when (< (count all-metadatas) (count table-dependents))
+      (log/warn "Some (possibly virtual) tables are not readable by the current user"))
+    all-metadatas))
+
+(defn- metadata-for-field-dependents
+  [field-dependents]
+  (let [metadatas (api.field/get-fields (map :id field-dependents))]
+    (when (< (count metadatas) (count field-dependents))
+      (log/warn "Some fields are not readable by the current user"))
+    (vec (sort-by :id metadatas))))
+
 (defn- metadata-for-dependents [dependents]
   {:tables (->> (:table dependents)
                 ;; Can be int or "card__<id>"
-                (sort-by (comp str :id))
-                (into []
-                      (keep #(fetch-or-warn
-                               %
-                               (fn [card-or-table-id]
-                                 (if-let [card-id (lib.util/legacy-string-table-id->card-id card-or-table-id)]
-                                   (api.table/fetch-card-query-metadata card-id)
-                                   (api.table/fetch-table-query-metadata card-or-table-id {})))))))
+                metadata-for-table-dependents)
    :databases (->> (:database dependents)
                    (sort-by :id)
                    (into []
                          (keep #(fetch-or-warn % api.database/get-database {}))))
    :fields (->> (:field dependents)
-                (sort-by :id)
-                (into []
-                      (keep #(fetch-or-warn % api.field/get-field {}))))})
+                metadata-for-field-dependents)})
+
+(defn- collect-source-tables
+  [query]
+  (let [from-joins (mapcat collect-source-tables (:joins query))]
+    (if-let [source-query (:source-query query)]
+      (concat (collect-source-tables source-query) from-joins)
+      (cond->> from-joins
+        (:source-table query) (cons (:source-table query))))))
```

Comment:
This is an optimization: collecting the things that will obviously be queried and put them into the metadata-provider in batches. If something is not there, it will be queried separately.